File: android/src/main/kotlin/com/ccpp/pgw_sdk/api/PaymentOptionDetailApi.kt
/*
 * Created by Nonthawatt Phongwittayapanu on 10/6/21, 3:39 PM
 * Copyright © 2021 2C2P. All rights reserved.
 */

package com.ccpp.pgw_sdk.api

import android.util.Log
import androidx.annotation.NonNull
import com.ccpp.pgw.sdk.android.callback.APIResponseCallback
import com.ccpp.pgw.sdk.android.core.PGWSDK
import com.ccpp.pgw.sdk.android.enums.APIResponseCode
import com.ccpp.pgw.sdk.android.enums.PaymentChannelCode
import com.ccpp.pgw.sdk.android.model.api.PaymentOptionDetailRequest
import com.ccpp.pgw.sdk.android.model.api.PaymentOptionDetailResponse
import com.ccpp.pgw.sdk.android.model.api.PaymentOptionRequest
import com.ccpp.pgw.sdk.android.model.api.PaymentOptionResponse
import com.ccpp.pgw_sdk.helper.JsonHelper
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class PaymentOptionDetailApi(
    @NonNull val call: MethodCall, @NonNull val result: MethodChannel.Result,
) {

    /**
     * Payment Option Detail
     */
    fun request() {
        val json = call.argument<String>("paymentOptionDetailRequest") ?: ""
        val request = paymentOptionDetailRequestFromJson(json) ?: kotlin.run {
            result.error("998", "Parse json error", null)
            return
        }

        PGWSDK.getInstance().paymentOptionDetail(request, object : APIResponseCallback<PaymentOptionDetailResponse> {
            override fun onResponse(response: PaymentOptionDetailResponse) {
                if (response.responseCode == APIResponseCode.APISuccess) {
                    //Read payment option detail response.
                    result.success(paymentOptionDetailResponseToJson(response))
                } else {
                    //Get error response and display error.
                    Log.d("DEMO", "onResponse: Error ${response.responseCode}")
                    Log.d("DEMO", "onResponse: Error ${response.responseDescription}")

                    result.error(response.responseCode, response.responseDescription, null)
                }
            }

            override fun onFailure(error: Throwable) {
                //Get error response and display error.
                Log.d("DEMO", "onFailure: $error")
                result.error("999", error.localizedMessage, error)
            }
        })
    }

    private fun paymentOptionDetailRequestFromJson(json: String): PaymentOptionDetailRequest? {
        val jsonObject = JSONObject(json)

        Log.d("DEMO", "parseJson: Request =>> ${jsonObject.toString(2)}")

        val paymentToken = JsonHelper.jsonString(jsonObject, "paymentToken")
        val clientId = JsonHelper.jsonString(jsonObject, "clientID")
        val locale = JsonHelper.jsonString(jsonObject, "locale")
        val category = JsonHelper.jsonString(jsonObject, "categoryCode")
        val group = JsonHelper.jsonString(jsonObject, "groupCode")

        return PaymentOptionDetailRequest(paymentToken).apply {
            if (!clientId.isNullOrBlank()) {
                setClientId(clientId)
            }
            if (!locale.isNullOrBlank()) {
                setLocale(locale)
            }
            categoryCode = category // PaymentChannelCode.Category.GlobalCardPayment;
            groupCode = group //PaymentChannelCode.Group.GlobalInstallmentPaymentPlan;
        }
    }

    private fun paymentOptionDetailResponseToJson(response: PaymentOptionDetailResponse): String {
        val channels: ArrayList<Map<String, Any?>> = arrayListOf()
        response.channels.forEach {
            channels.add(
                mapOf(
                    "sequenceNo" to it.sequenceNo,
                    "name" to it.name,
                    "iconUrl" to it.iconUrl,
                    "logoUrl" to it.logoUrl,
                )
            )
        }

        val cardTypes: ArrayList<Map<String, Any?>> = arrayListOf()
        response.validation.cardTypes.forEach {
            cardTypes.add(
                mapOf(
                    "sequenceNo" to it.sequenceNo,
                    "name" to it.name,
                    "iconUrl" to it.iconUrl,
                    "regex" to it.regex,
                    "prefixes" to it.prefixes,
                )
            )
        }

        val responseMap = mapOf(
            "name" to response.name,
            "categoryCode" to response.categoryCode,
            "groupCode" to response.groupCode,
            "iconUrl" to response.iconUrl,
            "validation" to mapOf(
                "cardNo" to mapOf(
                    "prefixes" to response.validation.cardNo.prefixes,
                    "regex" to response.validation.cardNo.regex,
                ),
                "cardTypes" to cardTypes,
            ),
            "channels" to channels,
            "totalChannel" to response.totalChannel,
            "configurationInfo" to mapOf(
                "payment" to mapOf(
                    "exchangeRate" to mapOf(
                        "mcp" to mapOf(
                            "terms" to response.configurationInfo.payment.exchangeRate.multipleCurrencyPricing.terms,
                        ),
                        "dcc" to mapOf(
                            "terms" to response.configurationInfo.payment.exchangeRate.dynamicCurrencyConversion.terms,
                        ),
                        "apmMcc" to mapOf(
                            "terms" to response.configurationInfo.payment.exchangeRate.alternativePaymentMethodMultipleCurrencyConversion.terms,
                        ),
                    ),
                ),
                "responseCode" to response.responseCode,
                "responseDescription" to response.responseDescription,
            ),
        )

        return JSONObject(responseMap).toString()
    }
}

File: ios/Classes/Api/PaymentOptionDetailApi.swift
//
//  PaymentOptionDetailApi.swift
//  pgw_sdk
//
//  Created by Nonthawatt Phongwittayapanu on 7/10/2564 BE.
//  Copyright © 2021 2C2P. All rights reserved.
//

import UIKit
import PGW

class PaymentOptionDetailApi {
    private var call: FlutterMethodCall
    private var result: FlutterResult
    
    init(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        self.call = call
        self.result = result
    }
    
    func request() {
        guard let dicts = call.arguments as? [String: Any],
              let jsonString = dicts["paymentOptionDetailRequest"] as? String
        else {
            result(FlutterError.init(code: "998", message: "Arguments exception", details: nil))
            return
        }
        guard let request = mappingRequest(json: jsonString) else {
            result(FlutterError.init(code: "900", message: "Json exception", details: nil))
            return
        }
        
        PGWSDK.shared.paymentOptionDetail(paymentOptionDetailRequest: request, { (response: PaymentOptionDetailResponse) in
            if response.responseCode == APIResponseCode.APISuccess {
                if let jsonString = self.mappingResposne(response) {
                    self.result(jsonString)
                } else {
                    self.result(FlutterError.init(code: response.responseCode, message: "Parse json error, \(response.responseDescription ?? "")", details: nil))
                }
            } else {
                //Get error response and display error.
                self.result(FlutterError.init(code: response.responseCode, message: response.responseDescription, details: nil))
            }
        }) { (error: NSError) in
            //Get error response and display error.
            self.result(FlutterError.init(code: "\(error.code)", message: error.localizedDescription, details: nil))
        }
    }
    
    // MARK: - Request
    private func mappingRequest(json: String) -> PaymentOptionDetailRequest? {
        do {
            let jsonData = json.data(using: .utf8)!
            let jsonDict = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any?]
            let paymentToken = jsonDict?["paymentToken"] as? String ?? ""
            
            let request = PaymentOptionDetailRequest(paymentToken: paymentToken)
            
            if let clientId = jsonDict?["clientID"] as? String {
                request.clientId = clientId
            }
            if let locale = jsonDict?["locale"] as? String {
                request.locale = locale
            }
            if let clientId = jsonDict?["categoryCode"] as? String {
                request.categoryCode = clientId
            }
            if let locale = jsonDict?["groupCode"] as? String {
                request.groupCode = locale
            }
            return request
        } catch let error as NSError {
            print(error)
            return nil
        }
    }
    
    // MARK: - Response
    private func mappingResposne(_ response: PaymentOptionDetailResponse) -> String? {
        var channels: [[String: Any?]] = []
        response.channels.forEach { (it: PaymentChannel) in
            channels.append(
                [
                    "sequenceNo": it.sequenceNo,
                    "name": it.name,
                    "iconUrl": it.iconUrl,
                    "logoUrl": it.logoUrl,
                ]
            )
        }
        
        var cardTypes: [[String: Any?]] = []
        response.validation.cardTypes.forEach { (it: CardTypeValidation) in
            cardTypes.append(
                [
                    "sequenceNo": it.sequenceNo,
                    "name": it.name,
                    "iconUrl": it.iconUrl,
                    "regex": it.regex,
                    "prefixes": it.prefixes,
                ]
            )
        }
        
        let responseDict: [String : Any?] = [
            "name": response.name,
            "categoryCode": response.categoryCode,
            "groupCode": response.groupCode,
            "iconUrl": response.iconUrl,
            "validation": [
                "cardNo": [
                    "prefixes": response.validation.cardNo.prefixes ?? [],
                    "regex": response.validation.cardNo.regex ?? "",
                ],
                "cardTypes": cardTypes,
            ],
            "channels": channels,
            "totalChannel": response.totalChannel,
            "configurationInfo": [
                "payment": [
                    "exchangeRate": [
                        "mcp": [
                            "terms": response.configurationInfo.payment.exchangeRate.multipleCurrencyPricing.terms,
                        ],
                        "dcc": [
                            "terms": response.configurationInfo.payment.exchangeRate.dynamicCurrencyConversion.terms,
                        ],
                        "apmMcc": [
                            "terms": response.configurationInfo.payment.exchangeRate.alternativePaymentMethodMultipleCurrencyConversion.terms,
                        ],
                    ],
                ],
                "configurationInfo":[
                    "payment":[
                        "exchangeRate":[
                            "mcp":[
                                "terms": response.configurationInfo.payment.exchangeRate.multipleCurrencyPricing.terms,
                            ],
                            "dcc":[
                                "terms": response.configurationInfo.payment.exchangeRate.dynamicCurrencyConversion.terms,
                            ],
                            "apmMcc":[
                                "terms": response.configurationInfo.payment.exchangeRate.alternativePaymentMethodMultipleCurrencyConversion.terms,
                            ],
                        ],
                    ],
                ],
            ],
            "responseCode": response.responseCode as String,
            "responseDescription": response.responseDescription as String,
        ]
        
        do {
            let data = try JSONSerialization.data(withJSONObject: responseDict)
            let string = String(data: data, encoding: .utf8)
            print(string ?? "")
            return string
        } catch let error as NSError {
            print(error)
            return nil
        }
    }
}
